### DISCUSSION TITLE
Mouse interaction with endless tilemap

### DISCUSSION THREAD

**Discussion by ZarimSF** [bevyengine/bevy#3256] [Q&A] [answered]
Hi,
I want to begin a bevy game where you have an 'endless' 2D tiled world.
#### Gameplay
Tiles should be able to 'do stuff'. By that I mean if you place, say, a factory on a tile, it would periodically produce some product and store it in a nearby depot.
#### Controls
There is no character the camera is centered on. Instead, you can move the camera around using arrow keys/wasd and interact with the tiles via the mouse. Perhaps hovering over a tile highlights it, and clicking it places or removes a building.
If you know games like [shapez.io](https://shapez.io/): that's basically what I want to achieve.
#### Questions
1. I found two plugins, [bevy_tilemap](https://github.com/joshuajbouw/bevy_tilemapl) and [bevy_ecs_tilemap](https://github.com/StarArawn/bevy_ecs_tilemap). Which one would be better suited for my needs?
2. How would I be able to interact with tiles via the cursor? I guess it's a bit naive to assume I could give a tile a `on_click()` function, right? Would I manually have to convert screen coordinates to world/tile coordinates or is there a smarter way?

I hope this is the right place to ask this question. 
Thanks in advance for your thoughts.

**Comment by wilk10** [comment, marked as answer]
Thanks for your question! Sounds like a fun idea for a game. 

Regarding your first questions, I haven't used either of the plugins, so maybe try to play around with both and see what fits better your needs. What i know is that `bevy_ecs_tilemap` has an `Entity` per tile, which means that it's very easy to attach specific behaviours to individual tiles, by inserting corresponding components. But maybe try both and decide based on that.

As per the second question, you'd need to map the cursor position to world space, see [in the cheatbook](https://bevy-cheatbook.github.io/cookbook/cursor2world.html) how it's done. Once you have the cursor coordinates in world space, in a system you can query for your Tiles, and determine if the cursor coordinates are over a specific Tile, based on the Tile's `Transform` and the size of the Tile, basically some flavour of [this system](https://github.com/bevyengine/bevy/blob/38c7d5eb9e81ab8e1aec03673599b25a9aa0c69c/crates/bevy_ui/src/focus.rs#L44) in Bevy, using to your components.

**Reply by ZarimSF** [reply]
Thank you very much, both code samples look very promising. I'll definitely have to fiddle around with the code, but that's also part of the fun :)
I'll mark this as answered because I think I got the help I needed to get start.
If there are more suggestions I will happily read them too.

**Reply by wilk10** [reply]
glad i could help! don't forget to showcase your creation in Discord (and/or ask for on-the-fly help)
